### PR TITLE
Add Custom Exceptions in Java Client

### PR DESCRIPTION
**Pull request by poojabhat0**
Add custom exceptions that map to each HTTP Status Code
Add statusCode as a variable for the custom exceptions and a getter for it

Related Issue: https://github.com/tecton-ai/tecton-http-client-java/issues/56

### CODE REVIEW

File: src/main/java/ai/tecton/client/exceptions/BadRequestException.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package ai.tecton.client.exceptions;
+
+/**
+ * An exception class representing a client error caused by a bad request. This exception is
+ * typically thrown when the client's request is malformed or contains invalid data, resulting in a
+ * "400 Bad Request" HTTP response status. It extends the {@link TectonClientException} class.
+ */
+public class BadRequestException extends TectonClientException {
+
+  public BadRequestException(String errorMessage, int statusCode) {
```
Comment (achals) [inline src/main/java/ai/tecton/client/exceptions/BadRequestException.java:10]: finals

File: src/main/java/ai/tecton/client/exceptions/ForbiddenException.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * An exception class representing a client error caused by forbidden access. This exception is
+ * typically thrown when the API Key in the request is authenticated, but the Service Account
+ * associated with the API Key is not authorized to access the workspace, resulting in a "403
+ * Forbidden" HTTP response status. It extends the {@link TectonClientException} class.
+ */
+public class ForbiddenException extends TectonClientException {
+
+  public ForbiddenException(String errorMessage, int statusCode) {
```
Comment (achals) [inline src/main/java/ai/tecton/client/exceptions/ForbiddenException.java:11]: finals

File: src/main/java/ai/tecton/client/exceptions/GatewayTimeoutExceptionException.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * An exception class representing a client error caused by a gateway timeout. This exception is
+ * typically thrown when a server acting as a gateway or proxy did not receive a timely response
+ * from the Feature Server, resulting in a "504 Gateway Timeout" HTTP response status. It extends
+ * the {@link TectonServiceException} class.
+ */
+public class GatewayTimeoutExceptionException extends TectonServiceException {
+
+  public GatewayTimeoutExceptionException(String errorMessage, int statusCode) {
```
Comment (achals) [inline src/main/java/ai/tecton/client/exceptions/GatewayTimeoutExceptionException.java:11]: finals

File: src/main/java/ai/tecton/client/exceptions/HttpStatusExceptionFactory.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    exceptionMappings.put(401, UnauthorizedException.class);
+    exceptionMappings.put(403, ForbiddenException.class);
+    exceptionMappings.put(404, ResourceNotFoundException.class);
+    exceptionMappings.put(429, ResourceExhaustedException.class);
+    exceptionMappings.put(500, InternalServerErrorException.class);
+    exceptionMappings.put(503, ServiceUnavailableException.class);
+    exceptionMappings.put(504, GatewayTimeoutExceptionException.class);
+  }
+
+  public static Optional<TectonException> createException(int statusCode, String errorMessage) {
```
Comment (achals) [inline src/main/java/ai/tecton/client/exceptions/HttpStatusExceptionFactory.java:23]: final

File: src/main/java/ai/tecton/client/exceptions/HttpStatusExceptionFactory.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    Class<? extends TectonException> exceptionClass = exceptionMappings.get(statusCode);
+    if (exceptionClass == null) {
+      return Optional.empty();
+    }
+    try {
+      return Optional.of(
+          exceptionClass
+              .getConstructor(String.class, int.class)
+              .newInstance(errorMessage, statusCode));
```
Comment (achals) [inline src/main/java/ai/tecton/client/exceptions/HttpStatusExceptionFactory.java:32]: eeeh.. Is using reflection going to incur a performance hit?

File: src/main/java/ai/tecton/client/exceptions/InternalServerErrorException.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package ai.tecton.client.exceptions;
+
+/**
+ * An exception class representing a server error caused by an internal server failure. This
+ * exception is typically thrown when an unexpected error occurs on the server, resulting in a "500
+ * Internal Server Error" HTTP response status. It extends the {@link TectonServiceException} class.
+ */
+public class InternalServerErrorException extends TectonServiceException {
+
+  public InternalServerErrorException(String errorMessage, int statusCode) {
```
Comment (achals) [inline src/main/java/ai/tecton/client/exceptions/InternalServerErrorException.java:10]: finals

File: src/main/java/ai/tecton/client/exceptions/TectonException.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.Optional;
+
+/** Base class that is used to represent various errors encountered by the Java client. */
+public class TectonException extends RuntimeException {
+  int statusCode;
+
+  public TectonException(String errorMessage) {
+    super(errorMessage);
+  }
```
[the remark this answers is not in the document]
Comment (poojabhat0) [reply src/main/java/ai/tecton/client/exceptions/TectonException.java:11]: Its an Optional and we'll return Optional.empty() if null

**Review by achals — APPROVED** [review]

**Pull request merged**